### PR TITLE
Add checks for Kerberos properties

### PR DESCRIPTION
**Pull request by akshatnair**
Ensure that we have all the properties set for Hive connector
when using Kerberos to access Hive metastore.

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java [hunk 1]
```diff
@@ -44,6 +45,28 @@
     public KerberosHiveMetastoreAuthentication(HiveClientConfig hiveClientConfig, @ForHiveMetastore HadoopAuthentication authentication)
     {
         this(hiveClientConfig.getHiveMetastoreServicePrincipal(), authentication);
+        verifyKerberosProperties(hiveClientConfig);
```
Comment (anusudarsan) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: I think we should be adding similar checks for configs in KerberosConfig.java. Those were the ones which had names similar to the airlift config properties.

Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: Okay that is from client to Presto server. I can add that.

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java [hunk 2]
```diff
@@ -44,6 +45,28 @@
...
+    }
+
+    private static void verifyKerberosProperties(HiveClientConfig hiveClientConfig)
+    {
+        requireNonNull(hiveClientConfig.getHiveMetastoreServicePrincipal(),
+                "Kerberos principal is not set for metastore service");
```
Comment (cawallin) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: These error messages should give the config property to be set, and be something like "hive.metastore.principal must be set when Kerberos is enabled for Hive."

File: presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java [hunk 3]
```diff
@@ -120,6 +122,16 @@ public SpnegoFilter(KerberosConfig config)
         }
     }
 
+    private static void verifyKerberosConfig(KerberosConfig config)
+    {
+        requireNonNull(config.getKerberosConfig(),
+                "Kerberos config file is not set");
```
Comment (cawallin) [inline presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java]: ditto on the error messages: something like "kerberos_config must be set when frontend Kerberos is enabled"

**Comment by cawallin** [comment]
The Travis failure in product tests seems relevant

File: presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java (hunk 3 above)
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java]: As per offline discussion we are going to keep this as it is.

**Comment by akshatnair** [comment]
@cawallin fixed the test failures. Build is passing now.

**Comment by akshatnair** [comment]
@martint please review

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java [hunk 4]
```diff
@@ -44,6 +45,28 @@
     public KerberosHiveMetastoreAuthentication(HiveClientConfig hiveClientConfig, @ForHiveMetastore HadoopAuthentication authentication)
     {
         this(hiveClientConfig.getHiveMetastoreServicePrincipal(), authentication);
+        verifyKerberosProperties(hiveClientConfig);
+    }
+
+    private static void verifyKerberosProperties(HiveClientConfig hiveClientConfig)
+    {
+        requireNonNull(hiveClientConfig.getHiveMetastoreServicePrincipal(),
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: No need to wrap these

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java [hunk 5]
```diff
@@ -44,6 +45,28 @@
...
+        requireNonNull(hiveClientConfig.getHiveMetastoreClientKeytab(),
+                "Kerberos keytab is not set for metastore client");
+
+        requireNonNull(hiveClientConfig.getHdfsPrestoPrincipal(),
+                "Kerberos principal is not set for HDFS");
+
+        requireNonNull(hiveClientConfig.getHdfsPrestoKeytab(),
+                "Kerberos keytab is not set for HDFS");
+
+        checkState(HdfsAuthenticationType.KERBEROS == hiveClientConfig.getHdfsAuthenticationType(),
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: This should be `checkArgument` since it's validating an argument. Also, flip the equality so the constant is on the left (which is the canonical way to write it).

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: We can include the invalid value here in the error message

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java [hunk 6]
```diff
@@ -44,6 +45,28 @@
     public KerberosHiveMetastoreAuthentication(HiveClientConfig hiveClientConfig, @ForHiveMetastore HadoopAuthentication authentication)
     {
         this(hiveClientConfig.getHiveMetastoreServicePrincipal(), authentication);
+        verifyKerberosProperties(hiveClientConfig);
+    }
+
+    private static void verifyKerberosProperties(HiveClientConfig hiveClientConfig)
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: Name the variable `config` so that the code is easier to read

File: presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java [hunk 7]
```diff
@@ -77,6 +78,7 @@
     @Inject
     public SpnegoFilter(KerberosConfig config)
     {
+        verifyKerberosConfig(config);
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java]: This is only used in one place, so inline the checks

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java (hunk 1 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: Inline since it's only used in one place

**Comment by electrum** [comment]
I'm confused about the purpose of this. Why aren't we checking the values where they are used?

**Comment by electrum** [comment]
Configs validation should normally be done using config validation. I think the issue here is that all the configs are in `HiveClientConfig`, so we can't make them `@NotNull` as that needs to be done conditionally.

We could solve this by refactoring the configs so they are in their own config class which is conditionally bound based on the type of authentication. This lets us use the normal validation which  prints the property name, etc.

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java (hunk 1 above)
Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: done

File: presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java (hunk 7 above)
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/server/security/SpnegoFilter.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java (hunk 6 above)
Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java (hunk 5 above)
Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java (hunk 4 above)
Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/authentication/KerberosHiveMetastoreAuthentication.java]: done

**Comment by akshatnair** [comment]
@electrum we wanted to ensure that once we know we are using the kerberos for authentication all the required fields are set. We had seen cases where one of the required property was not set and we it was very hard to debug it since it matched a property in http client code. To make it easier to identify such issues we are adding this check as soon as possible in the modules.

**Comment by electrum** [comment]
Why can't we validate the properties where they are used? It's very confusing to validate unused properties.

Also, can we restructure the configs to allow the normal config validation to work?

**Comment by electrum** [comment]
As a concrete example, look at `HiveSecurityModule`. It binds `LegacySecurityModule` only if `hive.security=legacy`. That module then binds `LegacySecurityConfig`. Thus, all those config properties are only allowed to be set in legacy security mode. It could also have required properties.

I think if we restructure the configs like that, we can do all the checks in the config system and additionally make the code easier to follow.

With the current approach, it's easy for someone to come along and delete all these validations because they appear to be unused (which sometimes happens when code is refactored and things are missed).

**Comment by akshatnair** [comment]
@electrum thanks for the feedback. Let me take a look at restructuring config option.

**Comment by akshatnair** [comment]
@electrum Sorry for the delay but I was out of office for a few days. I have addressed your comments. Please review it, thanks.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 8]
```diff
@@ -798,6 +755,7 @@ public HiveClientConfig setHdfsAuthenticationType(HdfsAuthenticationType hdfsAut
         return this;
     }
 
+    @NotNull
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Not needed for a primitive type (which can't be null)

File: presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java [hunk 9]
```diff
@@ -23,6 +25,7 @@
     private String serviceName;
     private File keytab;
 
+    @NotNull
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java]: These changes are unrelated to the Hive change. Can you pull into a separate commit or PR?

File: presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java [hunk 10]
```diff
@@ -47,6 +51,7 @@ public KerberosConfig setServiceName(String serviceName)
         return this;
     }
 
+    @NotNull
```
Comment (electrum) [inline presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java]: I'm not sure this is correct. There is a check for null in `SpnegoFilter` which makes setting the keytab optional. If we're going to require it, that check should be removed.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveKerberosConfig.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import io.airlift.configuration.Config;
+import io.airlift.configuration.ConfigDescription;
+
+import javax.validation.constraints.NotNull;
+
+public class HiveKerberosConfig
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveKerberosConfig.java]: Having these in the same file seems wrong, since it disallows configuring HDFS and Metastore independently. I think either of these configs will break:
```
hive.metastore.authentication.type=NONE
hive.hdfs.authentication.type=KERBEROS
```
or
```
hive.metastore.authentication.type=KERBEROS
hive.hdfs.authentication.type=NONE
```
These should probably be two separate configs, `HdfsKerberosConfig` and `MetastoreKerberosConfig`.

Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveKerberosConfig.java]: Currently we do not support only one of the options. It has to be enabled for both HDFS and Hive metastore or for none. That is why I have to keep them together.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java (hunk 8 above)
Comment (akshatnair) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: done

File: presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java (hunk 10 above)
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java]: Removed that check because we need to have the keytab if Kerberos is being used for authentication for CLI

File: presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java (hunk 9 above)
Comment (akshatnair) [inline presto-main/src/main/java/com/facebook/presto/server/security/KerberosConfig.java]: I will make a separate commit

**Comment by akshatnair** [comment]
@electrum I have addressed the comments. Build seems to be failing on master branch as well. I can wait for it to be fixed. Can you please review the changes in the meantime.

**Comment by electrum** [comment]
The second commit is actually for the server, not the CLI. The commit message should be something like "Make keytab required when configuring Kerberos authentication".

However, I'm still unclear on the motivation for this change. Per [documentation](https://docs.oracle.com/javase/8/docs/jre/api/security/jaas/spec/com/sun/security/auth/module/Krb5LoginModule.html), it seems reasonable to leave this optional:

> useKeyTab:
Set this to true if you want the module to get the principal's key from the the keytab.(default value is False) If keytab is not set then the module will locate the keytab from the Kerberos configuration file. If it is not specified in the Kerberos configuration file then it will look for the file {user.home}{file.separator}krb5.keytab.

Can we split out the two changes, and move the "make keytab required" change a separate PR? It seems to warrant further discussion. The `serviceName` `@NotNull` validation is a correct fix, since we'll get an NPE if it is not set.

**Comment by akshatnair** [comment]
Okay, I will move that commit out from this PR and we can discuss that requirement later.

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveKerberosConfig.java (hunk 11 above)
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveKerberosConfig.java]: If that's the case, then we should make them use a single option, so that the user can't misconfigure it. However, I'm wondering why they need to be tied together. Would the above actually work if we fixed the config code?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java [hunk 12]
```diff
@@ -100,14 +100,8 @@
...
     private HiveMetastoreAuthenticationType hiveMetastoreAuthenticationType = HiveMetastoreAuthenticationType.NONE;
-    private String hiveMetastoreServicePrincipal;
-    private String hiveMetastoreClientPrincipal;
-    private String hiveMetastoreClientKeytab;
-
     private HdfsAuthenticationType hdfsAuthenticationType = HdfsAuthenticationType.NONE;
-    private boolean hdfsImpersonationEnabled;
-    private String hdfsPrestoPrincipal;
-    private String hdfsPrestoKeytab;
+    private boolean hdfsImpersonationEnabled = false;
```
Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: This is unused

Comment (electrum) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientConfig.java]: Nevermind, I didn't notice the deletion above. This change is actually unnecessary, since it sets the field to the default Java value, which we don't do in our code style (IntelliJ has an inspection for this).

**Comment by electrum** [comment]
I updated your commit to split apart the two configs: https://github.com/prestodb/presto/pull/8161

Looking through the code, I don't see any reason why these can't be configured separately. They seem to be independent. What do you think?

**Comment by akshatnair** [comment]
They seems to be independent in code but from what I have read from our recommendations and documentation, we always want them to be configured together. I have asked the engineer who wrote that piece of code to confirm this. He is offline at the moment.
